metrics/honeycomb.go

```diff
@@ -283,75 +278,89 @@ func average(vals []float64) float64 {
 	switch metricType {
 	case "counter":
-		h.countersLock.Lock()
-		defer h.countersLock.Unlock()
-		// inside the lock, let's not race to create the counter
-		_, ok := h.counters[name]
-		if !ok {
-			newCounter := &counter{
-				name: name,
-			}
-			h.counters[name] = newCounter
-		}
+		getOrAdd(&h.lock, name, h.counters, createCounter)
 	case "gauge":
-		h.gaugesLock.Lock()
-		defer h.gaugesLock.Unlock()
-		_, ok := h.gauges[name]
-		if !ok {
-			newGauge := &gauge{
-				name: name,
-			}
-			h.gauges[name] = newGauge
-		}
+		getOrAdd(&h.lock, name, h.gauges, createGague)
 	case "histogram":
-		h.histogramsLock.Lock()
-		defer h.histogramsLock.Unlock()
-		_, ok := h.histograms[name]
-		if !ok {
-			newGauge := &histogram{
-				name: name,
-				vals: make([]float64, 0),
-			}
-			h.histograms[name] = newGauge
-		}
+		getOrAdd(&h.lock, name, h.histograms, createHistogram)
 	default:
 		h.Logger.Debug().Logf("unspported metric type %s", metricType)
 	}
 }
 
-func (h *HoneycombMetrics) Count(name string, n interface{}) {
-	count, ok := h.counters[name]
+// getOrAdd attempts to retrieve a (generic) metric from the provided map by name, wrapping the read operation
+// with a read lock (RLock). If the metric is not present in the map, it aquire a write lock and executes
+// create function and add it's to the map.
+func getOrAdd[T *counter | *gauge | *histogram](lock *sync.RWMutex, name string, metrics map[string]T, createMetric func(name string) T) T {
+	// attempt to get metric by name using read lock
+	lock.RLock()
+	metric, ok := metrics[name]
+	lock.RUnlock()
+
+	// if found, return existing metric
+	if ok {
+		return metric
+	}
+
+	// aquire write lock
+	lock.Lock()
+	// check again to see if it's been while waiting for write lock
+	metric, ok = metrics[name]
 	if !ok {
-		h.Register(name, "counter")
-		count = h.counters[name]
+		// create new metric using create function and add to map
+		metric := createMetric(name)
+		metrics[name] = metric
 	}
-	count.lock.Lock()
-	defer count.lock.Unlock()
-	count.val = count.val + int(ConvertNumeric(n))
+	lock.Unlock()
+	return metric
+}
+
+func createCounter(name string) *counter {
+	return &counter{
+		name: name,
+	}
+}
+
+func createGague(name string) *gauge {
+	return &gauge{
+		name: name,
+	}
+}
+
+func createHistogram(name string) *histogram {
+	return &histogram{
+		name: name,
+		vals: make([]float64, 0),
+	}
+}
+
+func (h *HoneycombMetrics) Count(name string, n interface{}) {
+	counter := getOrAdd(&h.lock, name, h.counters, createCounter)
+
+	// update value, using counter's lock
+	counter.lock.Lock()
+	counter.val = counter.val + int(ConvertNumeric(n))
+	counter.lock.Unlock()
 }
 
 func (h *HoneycombMetrics) Increment(name string) {
 	h.Count(name, 1)
 }
 
 func (h *HoneycombMetrics) Gauge(name string, val interface{}) {
-	gauge, ok := h.gauges[name]
-	if !ok {
-		h.Register(name, "gauge")
-		gauge = h.gauges[name]
-	}
+	gauge := getOrAdd(&h.lock, name, h.gauges, createGague)
```

Comment:
```suggestion
	gauge := getOrAdd(&h.lock, name, h.gauges, createGauge)
```
(with function rename)